### PR TITLE
Add Visdom-based training diagnostics logger

### PR DESCRIPTION
**Pull request by Sahilkatkamwar**
## Summary
This PR introduces a Visdom-based training diagnostics logger to improve visibility into model training.

## What it adds
- Batched scalar logging for training metrics
- Automatic basic metric discovery
- Learning-rate logging for optimizer param groups
- Gradient anomaly detection for NaN, exploding, and dead gradients
- Optional model-health summaries
- Optional weight histograms and heatmaps
- Custom metric registration
- Safe attach/detach lifecycle for PyTorch models

## Notes
The logger is designed to be lightweight by default, with heavier diagnostics kept optional and periodic so training stays responsive.

For detailed usage, examples, and screenshots of the dashboard, please refer to the README.

## Summary by Sourcery

Introduce a Visdom-based PyTorch training diagnostics logger and enhance the Visdom UI with richer export controls and history.

New Features:
- Add a PyTorch-native VisdomDiagnosticsLogger that batches scalar logging, auto-discovers basic metrics, tracks learning rates, and supports custom metrics and model summaries.
- Provide gradient anomaly detection and model health visualizations including heatmaps and parameter histograms for PyTorch models.
- Expose helper utilities for gradient analysis, metric computation, and parameter hooks under a new visdom.integrations.pytorch package.
- Add an MNIST training example demonstrating how to use the diagnostics logger in a real training loop.
- Extend the plot pane UI with an export format menu, per-pane export history view, and local export logging.

Enhancements:
- Disable Plotly's default image export button in favor of the custom export menu and ensure plots resize correctly when toggling export history.

Documentation:
- Add high-level integration entry points and example script to serve as documentation for using Visdom with PyTorch diagnostics.

Tests:
- Add unit tests for the PyTorch Visdom logger covering anomaly warnings, automatic metric logging, custom metric registration, and gradient anomaly hooks.

## Documentation

For detailed usage, examples, and dashboard visuals:

👉 [View Logger README](https://github.com/Sahilkatkamwar/visdom/blob/logger-feature/py/README.md)

### CODE REVIEW

File: js/panes/Pane.js [hunk 1]
```diff
@@ -6,19 +6,20 @@
  * LICENSE file in the root directory of this source tree.
  *
  */
-
+//Pane.js
```
Comment (rajnisht7) [inline js/panes/Pane.js:9]: whats the point of writing this

File: py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+            logger.auto_log(
+                outputs=outputs.detach(),
+                targets=targets.detach(),
+                loss=loss.detach(),
+                step=global_step,
+                group="train",
+            )
+            
+            # 🔥 baseline (ALWAYS log 0)
```
Comment (rajnisht7) [inline py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py:83]: why there is emoji in your comment
or LLM

File: py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            logger.log_warning("exploding_grad", 0.0, step=global_step)
+            logger.log_warning("nan_grad", 0.0, step=global_step)
+
+            # 🔥 spikes
```
Comment (rajnisht7) [inline py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py:87]: again emoji in comment

File: tests/test_pytorch_logger.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    logger.register_metric("f1_score", lambda outputs, targets: 0.5, group="custom")
+
+    outputs = torch.tensor([[2.0, 1.0], [0.1, 0.9]])
+    targets = torch.tensor([0, 1])
+
+    logger.auto_log(outputs=outputs, targets=targets, step=2, group="train")
+    logger.flush()
+
+    titles = [call["opts"]["title"] for call in dummy.calls if call["method"] == "line"]
+    assert "custom/f1_score" in titles
```
[the remark this answers is not in the document]
Comment (Sahilkatkamwar) [reply tests/test_pytorch_logger.py:154]: implemented PyTorch logger with tests and add smooth widget toggle in PlotPane
Incorporates improvements suggested by Sourcery AI✅

File: py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py (hunk 2 above)
Comment (Sahilkatkamwar) [reply py/visdom/integrations/pytorch/examples/train_mnist_diagnostics.py:83]: The emoji is just for visual emphasis to highlight the baseline section. It has no effect on the code and is fully valid in Python comments.

File: js/panes/Pane.js (hunk 1 above)
Comment (Sahilkatkamwar) [reply js/panes/Pane.js:9]: This is included to indicate the target file for anyone copying the snippet outside the repository context.

**Review by rajnisht7 — COMMENTED** [review]
there is already readme.md therefore no need of an explicit `readme.md` file

**Comment by rajnisht7** [comment]
@Sahilkatkamwar kindly address the copilot reviews as well

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@Sahilkatkamwar Can u please go through the copilot reviews again and address all high priority and medium priority issues and also add proper video/ screenshots of impact of ur changes on visdom .

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@Sahilkatkamwar There are a few blocking issues that need to be fixed . The main problems are the incorrect .gitignore entries, the close() crash caused by self.flush Please go through it also

**Comment by mariobehling** [comment]
Thanks for the contribution.

We are closing this PR because it is not ready to be merged in its current form. The PR introduces a large new PyTorch diagnostics logger, additional integration utilities, tests, an MNIST example, and also changes the frontend export UI. This is a broad scope and mixes several areas that should be handled separately.

There are also unresolved blocking review points, including:

* Copilot and reviewer comments have not been fully addressed
* High and medium priority review issues still need follow-up
* The `close()` / `self.flush` crash needs to be fixed
* Incorrect `.gitignore` entries need to be corrected
* The behavior and impact of the UI changes need proper screenshots or a video
* The frontend export changes should not be bundled together with a new PyTorch diagnostics logger unless there is a clear technical reason

For a large feature like this, we need a smaller, well-scoped PR that is easier to review and maintain. A new PR would be welcome if it:

* Focuses only on the PyTorch diagnostics logger, or only on the export UI improvements
* Addresses all blocking review comments
* Includes clear documentation and usage examples
* Provides tests for the implemented behavior
* Adds screenshots or a video where UI behavior changes
* Keeps unrelated changes out of the PR

Closing this one for now. Thanks again for working on it.

**Comment by Sahilkatkamwar** [comment]
I apologise for the inconvenience , definitely would be giving it another
try for a smaller scope of PR better for both review and merge.
